File: star-proto/tests/go/motor_control_test.go
// Package starproto_test provides serialization and functionality tests for STAR Protocol Buffer messages.
// This file tests motor control commands including velocity commands for 4-motor differential drive.
package starproto_test

import (
	"testing"

	starv1 "github.com/Locked-Inc/star-proto/gen/go/star/v1"
	"google.golang.org/protobuf/proto"
)

// TestVelocityCommand_4Motors verifies that VelocityCommand supports 4 independent motors
func TestVelocityCommand_4Motors(t *testing.T) {
	cmd := &starv1.VelocityCommand{
		FrontLeftVelocityMps:  1.5,
		BackLeftVelocityMps:   1.5,
		FrontRightVelocityMps: 1.0,
		BackRightVelocityMps:  1.0,
		Sequence:              42,
		TimestampUs:           1000000,
	}

	// Serialize
	data, err := proto.Marshal(cmd)
	if err != nil {
		t.Fatalf("Failed to marshal VelocityCommand: %v", err)
	}

	// Verify compact encoding (4 motors should still be small)
	if len(data) > 100 {
		t.Errorf("Encoded size %d bytes is too large (expected < 100)", len(data))
	}

	// Deserialize
	decoded := &starv1.VelocityCommand{}
	if err := proto.Unmarshal(data, decoded); err != nil {
		t.Fatalf("Failed to unmarshal VelocityCommand: %v", err)
	}

	// Verify all 4 motor fields
	if decoded.FrontLeftVelocityMps != 1.5 {
		t.Errorf("Front left velocity = %f, want 1.5", decoded.FrontLeftVelocityMps)
	}
	if decoded.BackLeftVelocityMps != 1.5 {
		t.Errorf("Back left velocity = %f, want 1.5", decoded.BackLeftVelocityMps)
	}
	if decoded.FrontRightVelocityMps != 1.0 {
		t.Errorf("Front right velocity = %f, want 1.0", decoded.FrontRightVelocityMps)
	}
	if decoded.BackRightVelocityMps != 1.0 {
		t.Errorf("Back right velocity = %f, want 1.0", decoded.BackRightVelocityMps)
	}
	if decoded.Sequence != 42 {
		t.Errorf("Sequence = %d, want 42", decoded.Sequence)
	}
	if decoded.TimestampUs != 1000000 {
		t.Errorf("TimestampUs = %d, want 1000000", decoded.TimestampUs)
	}
}

// TestVelocityCommand_DifferentialDrive tests differential drive mode
// where front/back left = left side, front/back right = right side
func TestVelocityCommand_DifferentialDrive(t *testing.T) {
	leftVel := 1.5
	rightVel := 1.0

	cmd := &starv1.VelocityCommand{
		FrontLeftVelocityMps:  leftVel,  // Left front
		BackLeftVelocityMps:   leftVel,  // Left rear
		FrontRightVelocityMps: rightVel, // Right front
		BackRightVelocityMps:  rightVel, // Right rear
		Sequence:              1,
	}

	// Serialize and deserialize
	data, err := proto.Marshal(cmd)
	if err != nil {
		t.Fatalf("Marshal failed: %v", err)
	}

	decoded := &starv1.VelocityCommand{}
	if err := proto.Unmarshal(data, decoded); err != nil {
		t.Fatalf("Unmarshal failed: %v", err)
	}

	// Verify differential drive pairing
	if decoded.FrontLeftVelocityMps != decoded.BackLeftVelocityMps {
		t.Error("Left motors (front & back) should have same velocity")
	}
	if decoded.FrontRightVelocityMps != decoded.BackRightVelocityMps {
		t.Error("Right motors (front & back) should have same velocity")
	}
}

// TestVelocityCommand_VelocityRange tests valid velocity range (±2.0 m/s)
func TestVelocityCommand_VelocityRange(t *testing.T) {
	testCases := []struct {
		name     string
		velocity float64
		valid    bool
	}{
		{"Zero velocity", 0.0, true},
		{"Max forward", 2.0, true},
		{"Max reverse", -2.0, true},
		{"Within range positive", 1.5, true},
		{"Within range negative", -1.5, true},
		{"Exceeds max", 2.5, false},
		{"Exceeds min", -2.5, false},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			cmd := &starv1.VelocityCommand{
				FrontLeftVelocityMps:  tc.velocity,
				BackLeftVelocityMps:   tc.velocity,
				FrontRightVelocityMps: tc.velocity,
				BackRightVelocityMps:  tc.velocity,
			}

			data, err := proto.Marshal(cmd)
			if err != nil {
				t.Fatalf("Marshal failed: %v", err)
			}

			decoded := &starv1.VelocityCommand{}
			if err := proto.Unmarshal(data, decoded); err != nil {
				t.Fatalf("Unmarshal failed: %v", err)
			}

			// Protocol buffers will serialize any value, but application
			// should validate and clamp to ±2.0 m/s range
			if tc.valid {
				if decoded.FrontLeftVelocityMps != tc.velocity {
					t.Errorf("Velocity = %f, want %f", decoded.FrontLeftVelocityMps, tc.velocity)
				}
			}
		})
	}
}

// TestEncoderData_4Motors verifies EncoderData supports motor IDs 0-3
func TestEncoderData_4Motors(t *testing.T) {
	testCases := []struct {
		motorID     int32
		ticks       int64
		velocityMps float64
	}{
		{0, 1000, 1.5},
		{1, 2000, 1.2},
		{2, 1500, 1.0},
		{3, 1800, 0.8},
	}

	for _, tc := range testCases {
		encoder := &starv1.EncoderData{
			MotorId:     tc.motorID,
			Ticks:       tc.ticks,
			VelocityMps: tc.velocityMps,
			TimestampUs: 1000000,
		}

		// Serialize
		data, err := proto.Marshal(encoder)
		if err != nil {
			t.Fatalf("Motor %d: Marshal failed: %v", tc.motorID, err)
		}

		// Deserialize
		decoded := &starv1.EncoderData{}
		if err := proto.Unmarshal(data, decoded); err != nil {
			t.Fatalf("Motor %d: Unmarshal failed: %v", tc.motorID, err)
		}

		// Verify fields
		if decoded.MotorId != tc.motorID {
			t.Errorf("MotorId = %d, want %d", decoded.MotorId, tc.motorID)
		}
		if decoded.Ticks != tc.ticks {
			t.Errorf("Ticks = %d, want %d", decoded.Ticks, tc.ticks)
		}
		if decoded.VelocityMps != tc.velocityMps {
			t.Errorf("VelocityMps = %f, want %f", decoded.VelocityMps, tc.velocityMps)
		}
	}
}

// TestMotorStatus_4Motors verifies MotorStatus supports motor IDs 0-3
func TestMotorStatus_4Motors(t *testing.T) {
	t.Skip("MotorStatus message removed - unused in firmware")
}

// TestMotorPowerCommand_4Motors verifies direct motor power control
func TestMotorPowerCommand_4Motors(t *testing.T) {
	t.Skip("MotorPowerCommand message removed - unused in firmware")
}

// TestSetVelocityRequest_Roundtrip tests complete request/response flow
func TestSetVelocityRequest_Roundtrip(t *testing.T) {
	req := &starv1.SetVelocityRequest{
		Header: &starv1.RequestHeader{
			RequestId: "test-request-123",
		},
		Command: &starv1.VelocityCommand{
			FrontLeftVelocityMps:  1.5,
			BackLeftVelocityMps:   1.5,
			FrontRightVelocityMps: 1.0,
			BackRightVelocityMps:  1.0,
			Sequence:              1,
			TimestampUs:           1000000,
		},
	}

	data, err := proto.Marshal(req)
	if err != nil {
		t.Fatalf("Marshal failed: %v", err)
	}

	decoded := &starv1.SetVelocityRequest{}
	if err := proto.Unmarshal(data, decoded); err != nil {
		t.Fatalf("Unmarshal failed: %v", err)
	}

	if decoded.Header.RequestId != "test-request-123" {
		t.Errorf("RequestId = %s, want test-request-123", decoded.Header.RequestId)
	}
	if decoded.Command.FrontLeftVelocityMps != 1.5 {
		t.Errorf("Front left velocity = %f, want 1.5", decoded.Command.FrontLeftVelocityMps)
	}
}

// TestEmergencyStopRequest verifies E-STOP request encoding
func TestEmergencyStopRequest(t *testing.T) {
	req := &starv1.EmergencyStopRequest{
		Header: &starv1.RequestHeader{
			RequestId: "estop-request-1",
		},
		Reason: "User initiated emergency stop",
	}

	data, err := proto.Marshal(req)
	if err != nil {
		t.Fatalf("Marshal failed: %v", err)
	}

	decoded := &starv1.EmergencyStopRequest{}
	if err := proto.Unmarshal(data, decoded); err != nil {
		t.Fatalf("Unmarshal failed: %v", err)
	}

	if decoded.Reason != "User initiated emergency stop" {
		t.Errorf("Reason = %s, want 'User initiated emergency stop'", decoded.Reason)
	}
}

File: star-proto/tests/typescript/src/motor_control.test.ts
// motor_control.test.ts - TypeScript Motor Control Tests
// Verifies 4-motor control schema serialization and validation.
//
// STAR Project - Texas A&M University
// January 2026

import { describe, it, expect } from 'vitest';

// Import generated classes (will be available after buf generate)
// import { VelocityCommand, EncoderData, MotorStatus, MotorState } from '@gen/star/v1/motor_control';
// import { SetVelocityRequest, SetVelocityResponse } from '@gen/star/v1/motor_control';
// import { RequestHeader, ResponseHeader, Status } from '@gen/star/v1/common';

describe('Motor Control Tests', () => {

    describe('Test Framework', () => {
        it('should have working test framework', () => {
            // Placeholder test - verifies Vitest is operational
            expect(true).toBe(true);
        });
    });

    // TODO: Uncomment after buf generate produces TypeScript code
    /*
    describe('VelocityCommand - 4 Motors', () => {
        it('should support 4 independent motors', () => {
            const cmd: VelocityCommand = {
                motor0VelocityMps: 1.5,
                motor1VelocityMps: 1.5,
                motor2VelocityMps: 1.0,
                motor3VelocityMps: 1.0,
                sequence: 42,
                timestampUs: BigInt(1000000),
            };

            // Serialize
            const bytes = VelocityCommand.toBinary(cmd);

            // Verify compact encoding (4 motors should still be small)
            expect(bytes.length).toBeLessThan(100);

            // Deserialize
            const decoded = VelocityCommand.fromBinary(bytes);

            // Verify all 4 motor fields
            expect(decoded.motor0VelocityMps).toBeCloseTo(1.5, 3);
            expect(decoded.motor1VelocityMps).toBeCloseTo(1.5, 3);
            expect(decoded.motor2VelocityMps).toBeCloseTo(1.0, 3);
            expect(decoded.motor3VelocityMps).toBeCloseTo(1.0, 3);
            expect(decoded.sequence).toBe(42);
        });

        it('should support differential drive mode', () => {
            const leftVel = 1.5;
            const rightVel = 1.0;

            const cmd: VelocityCommand = {
                motor0VelocityMps: leftVel,  // Left front
                motor1VelocityMps: leftVel,  // Left rear
                motor2VelocityMps: rightVel, // Right front
                motor3VelocityMps: rightVel, // Right rear
                sequence: 1,
                timestampUs: BigInt(0),
            };

            const bytes = VelocityCommand.toBinary(cmd);
            const decoded = VelocityCommand.fromBinary(bytes);

            // Verify left motors match (0 & 1)
            expect(decoded.motor0VelocityMps).toBeCloseTo(leftVel, 3);
            expect(decoded.motor1VelocityMps).toBeCloseTo(leftVel, 3);

            // Verify right motors match (2 & 3)
            expect(decoded.motor2VelocityMps).toBeCloseTo(rightVel, 3);
            expect(decoded.motor3VelocityMps).toBeCloseTo(rightVel, 3);
        });

        it('should validate velocity range (±2.0 m/s)', () => {
            const testCases = [
                { name: 'zero', vel: 0.0, valid: true },
                { name: 'max forward', vel: 2.0, valid: true },
                { name: 'max reverse', vel: -2.0, valid: true },
                { name: 'within range', vel: 1.5, valid: true },
                { name: 'exceeds max', vel: 2.5, valid: false },
                { name: 'exceeds min', vel: -2.5, valid: false },
            ];

            testCases.forEach(tc => {
                const cmd: VelocityCommand = {
                    motor0VelocityMps: tc.vel,
                    motor1VelocityMps: tc.vel,
                    motor2VelocityMps: tc.vel,
                    motor3VelocityMps: tc.vel,
                    sequence: 0,
                    timestampUs: BigInt(0),
                };

                const bytes = VelocityCommand.toBinary(cmd);
                const decoded = VelocityCommand.fromBinary(bytes);

                // Protocol buffers will serialize any value, but application
                // should validate and clamp to ±2.0 m/s range
                if (tc.valid) {
                    expect(decoded.motor0VelocityMps).toBeCloseTo(tc.vel, 3);
                }
            });
        });

        it('should serialize to JSON correctly', () => {
            const cmd: VelocityCommand = {
                motor0VelocityMps: 1.5,
                motor1VelocityMps: 1.5,
                motor2VelocityMps: 1.0,
                motor3VelocityMps: 1.0,
                sequence: 42,
                timestampUs: BigInt(1000000),
            };

            const json = VelocityCommand.toJsonString(cmd);
            const parsed = VelocityCommand.fromJsonString(json);

            expect(parsed.motor0VelocityMps).toBeCloseTo(1.5, 3);
            expect(parsed.motor1VelocityMps).toBeCloseTo(1.5, 3);
        });
    });

    describe('EncoderData - 4 Motors', () => {
        it('should support motor IDs 0-3', () => {
            const testCases = [
                { motorId: 0, ticks: 1000, velocityMps: 1.5 },
                { motorId: 1, ticks: 2000, velocityMps: 1.2 },
                { motorId: 2, ticks: 1500, velocityMps: 1.0 },
                { motorId: 3, ticks: 1800, velocityMps: 0.8 },
            ];

            testCases.forEach(tc => {
                const encoder: EncoderData = {
                    motorId: tc.motorId,
                    ticks: BigInt(tc.ticks),
                    velocityMps: tc.velocityMps,
                    timestampUs: BigInt(1000000),
                };

                const bytes = EncoderData.toBinary(encoder);
                const decoded = EncoderData.fromBinary(bytes);

                expect(decoded.motorId).toBe(tc.motorId);
                expect(Number(decoded.ticks)).toBe(tc.ticks);
                expect(decoded.velocityMps).toBeCloseTo(tc.velocityMps, 3);
            });
        });
    });

    describe('MotorStatus - 4 Motors', () => {
        it('should support motor IDs 0-3', () => {
            for (let motorId = 0; motorId < 4; motorId++) {
                const status: MotorStatus = {
                    motorId: motorId,
                    dutyCyclePercent: 75.5,
                    velocityMps: 1.5,
                    targetVelocityMps: 1.5,
                    temperatureCelsius: 45.0,
                    currentMa: 2500.0,
                    faultFlags: 0,
                    state: MotorState.MOTOR_STATE_RUNNING,
                };

                const bytes = MotorStatus.toBinary(status);
                const decoded = MotorStatus.fromBinary(bytes);

                expect(decoded.motorId).toBe(motorId);
                expect(decoded.dutyCyclePercent).toBeCloseTo(75.5, 1);
                expect(decoded.state).toBe(MotorState.MOTOR_STATE_RUNNING);
            }
        });
    });

    describe('SetVelocityRequest/Response', () => {
        it('should complete request/response roundtrip', () => {
            const req: SetVelocityRequest = {
                header: {
                    requestId: 'test-request-123',
                    clientVersion: '1.0.0',
                },
                command: {
                    motor0VelocityMps: 1.5,
                    motor1VelocityMps: 1.5,
                    motor2VelocityMps: 1.0,
                    motor3VelocityMps: 1.0,
                    sequence: 1,
                    timestampUs: BigInt(1000000),
                },
            };

            const reqBytes = SetVelocityRequest.toBinary(req);
            const parsedReq = SetVelocityRequest.fromBinary(reqBytes);

            expect(parsedReq.header?.requestId).toBe('test-request-123');
            expect(parsedReq.command?.motor0VelocityMps).toBeCloseTo(1.5, 3);

            // Response
            const resp: SetVelocityResponse = {
                header: {
                    requestId: 'test-request-123',
                    status: Status.STATUS_OK,
                    errorMessage: '',
                    latencyUs: BigInt(0),
                },
            };

            const respBytes = SetVelocityResponse.toBinary(resp);
            const parsedResp = SetVelocityResponse.fromBinary(respBytes);

            expect(parsedResp.header?.requestId).toBe('test-request-123');
            expect(parsedResp.header?.status).toBe(Status.STATUS_OK);
        });
    });
    */
});
